sdk/servicebus/Azure.Messaging.ServiceBus/src/Amqp/BodyMemory.cs

```diff
@@ -0,0 +1,88 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections;
+using System.Collections.Generic;
+using Azure.Core;
+using Microsoft.Azure.Amqp.Encoding;
+using Microsoft.Azure.Amqp.Framing;
+
+namespace Azure.Messaging.ServiceBus.Amqp
+{
+    internal sealed class BodyMemory : IEnumerable<ReadOnlyMemory<byte>>
+    {
+        private ArrayBufferWriter<byte> _writer;
+        private IList<ReadOnlyMemory<byte>> _segments;
+        private readonly IEnumerable<ReadOnlyMemory<byte>> _lazySegments;
+
+        private ReadOnlyMemory<byte> WrittenMemory
+        {
+            get
+            {
+                if (_lazySegments != null)
+                {
+                    foreach (var segment in _lazySegments)
+                    {
+                        Append(segment);
+                    }
+                }
+
+                return _writer?.WrittenMemory ?? ReadOnlyMemory<byte>.Empty;
+            }
+        }
+
+        // for the send path the combined memory for the body is not precomputed
+        public BodyMemory(IEnumerable<ReadOnlyMemory<byte>> dataSegments)
+        {
+            _lazySegments = dataSegments;
+        }
+
+        // for the receive path the combined memory is precomputed because we need to copy the underlying AMQP data anyway
+        public BodyMemory(IEnumerable<Data> dataSegments)
+        {
+            foreach (var segment in dataSegments)
+            {
+                Append(segment);
+            }
+        }
+
+        public IEnumerator<ReadOnlyMemory<byte>> GetEnumerator()
+        {
+            return _segments?.GetEnumerator() ?? _lazySegments.GetEnumerator();
+        }
+
+        IEnumerator IEnumerable.GetEnumerator()
+        {
+            return GetEnumerator();
+        }
+
+        private void Append(ReadOnlyMemory<byte> segment)
+        {
+            _writer ??= new ArrayBufferWriter<byte>();
+            _segments ??= new List<ReadOnlyMemory<byte>>();
+
+            var memory = _writer.GetMemory(segment.Length);
+            segment.CopyTo(memory);
+            _writer.Advance(segment.Length);
+            _segments.Add(memory.Slice(0, segment.Length));
+        }
+
+        private void Append(DescribedType segment)
```

Comment:
Why is this needed?

Reply:
On the receive path we transform AMQP data types into ROM. I just picked the base type with is `DescribedType` instead of `Data`

Reply:
Any reason to not use `Data` since that is the type the AmqpMessage.Create expects?

Reply:
I changed it back